Source/NSFontCollection.m

```diff
@@ -0,0 +1,645 @@
+                           visibility: (NSFontCollectionVisibility)visibility
+                               toName: (NSFontCollectionName)name
+                                error: (NSError **)error
+{
+  NSFontCollection *collection = [_availableFontCollections objectForKey: aname];
+  BOOL rv = [collection _removeFile];
+
+  if (rv == YES)
+    {
+      [collection _setName: name];
+      [collection _writeToFile];
+    }
+  [NSFontCollection _loadAvailableFontCollections];
+
+  return rv;
+}
+
++ (NSArray *) allFontCollectionNames
+{
+  return [_availableFontCollections allKeys];
+}
+
++ (NSFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+{
+  NSFontCollection *fc = [_availableFontCollections objectForKey: name];
+  if (fc == nil)
+    {
+      // Getting named collections should not create new collections on disk.
+    }
+  return fc;
+}
+
++ (NSFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+                                   visibility: (NSFontCollectionVisibility)visibility
+{
+  return [self fontCollectionWithName: name];
+}
+
+// Descriptors
+- (NSArray *) queryDescriptors 
+{
+  return [_fontCollectionDictionary objectForKey: @"NSFontDescriptorAttributes"];
+}
+
+- (NSArray *) exclusionDescriptors
+{
+  return [_fontCollectionDictionary objectForKey: @"NSFontExclusionDescriptorAttributes"];
+}
+
+- (NSArray *) matchingDescriptors
+{
+  return [self matchingDescriptorsWithOptions: nil];
+}
+
+- (NSArray *) matchingDescriptorsWithOptions: (NSDictionary *)options
+{
+  GSFontEnumerator *fen = [GSFontEnumerator sharedEnumerator];
+  return [fen matchingFontDescriptorsFor: options];
+}
+
+- (NSArray *) matchingDescriptorsForFamily: (NSString *)family
+{
+  return [self matchingDescriptorsForFamily: family options: nil];
+}
+
+- (NSArray *) matchingDescriptorsForFamily: (NSString *)family options: (NSDictionary *)options
+{
+  NSMutableArray *r = [NSMutableArray arrayWithCapacity: 50];
+  NSArray *a = [self matchingDescriptorsWithOptions: options];
+  NSEnumerator *en = [a objectEnumerator];
+  id o = nil;
+
+  while((o = [en nextObject]) != nil)
+    {
+      if ([[o familyName] isEqualToString: family])
+        {
+          [r addObject: o];
+        }    
+    }
+  
+  return [a copy];
+}
+
+- (instancetype) copyWithZone: (NSZone *)zone
+{
+  NSFontCollection *fc = [[NSFontCollection allocWithZone: zone] init];
+  ASSIGNCOPY(fc->_fontCollectionDictionary, _fontCollectionDictionary);
+  return fc;
+}
+
+- (instancetype) mutableCopyWithZone: (NSZone *)zone
+{
+  NSMutableFontCollection *fc = [[NSMutableFontCollection allocWithZone: zone] init];
+
+  [fc _setFontCollectionDictionary: _fontCollectionDictionary];
+
+  return fc;
+}
+
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  self = [super init];
+  if (self != nil)
+    {
+      if ([coder allowsKeyedCoding])
+        {
+          [coder encodeObject: _fontCollectionDictionary
```

Comment:
You seem to have mixed the code between initWithCoder and encodeWithCoder. Just switch the implementation